File: aider_api.py
from fastapi import FastAPI, BackgroundTasks, HTTPException, Request
from fastapi.responses import JSONResponse, StreamingResponse
from sse_starlette.sse import EventSourceResponse
from pydantic import BaseModel
import subprocess
import asyncio
import sys
import os
from typing import Optional
import tempfile
import shutil
import logging
import shlex
import json

# Configure logging
logging.basicConfig(
    level=logging.DEBUG,
    filename='aider_api.log',  # Log to file instead of stdout
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

app = FastAPI(title="Aider API Server")

class AiderRequest(BaseModel):
    message: str
    files: Optional[dict[str, str]] = None  # filename: content, optional
    auto_commits: bool = True
    dirty_commits: bool = True
    dry_run: bool = False
    root: str = "."  # Directory to run aider in, defaults to current directory
    stream: bool = True  # Whether to stream the response using SSE

async def collect_aider_output(message: str, files: Optional[dict[str, str]], auto_commits: bool, 
                            dirty_commits: bool, dry_run: bool, root: str = ".", stream: bool = True):
    
    logger.debug(f"Starting stream_aider_output with message: {message}")
    logger.debug(f"Files to process: {list(files.keys()) if files else 'None'}")
    # Create a temporary directory to store the files if needed
    with tempfile.TemporaryDirectory() as temp_dir:
        # Write the files to the temporary directory if provided
        if files:
            for filename, content in files.items():
                file_path = os.path.join(temp_dir, filename)
                with open(file_path, 'w') as f:
                    f.write(content)

        # Get the aider executable from the venv's bin directory
        venv_dir = os.path.dirname(os.path.dirname(sys.executable))
        aider_path = os.path.join(venv_dir, 'bin', 'aider')
        
        # Prepare the aider command
        cmd = [
            aider_path,
            "--message", message, 
            # "--stream" if not dry_run else "--no-stream",
            "--auto-commits" if auto_commits else "--no-auto-commits",
            "--dirty-commits" if dirty_commits else "--no-dirty-commits",
            "--dry-run" if dry_run else "--no-dry-run", 
            "--no-show-model-warnings", # Do not show model warnings
            "--yes"  # Automatically confirm any prompts
        ]
        
        # Add model if specified in environment
        # if "AIDER_MODEL" in os.environ:
        #     cmd.extend(["--model", f'"{os.environ["AIDER_MODEL"]}"'])
            
        # Add files if provided
        if files:
            cmd.extend([os.path.join(temp_dir, f) for f in files.keys()])

        logger.debug(f"Working directory: {os.path.abspath(root)}")
        logger.debug(f"Using aider from: {aider_path}")
        logger.debug(f"Executing command: {' '.join(cmd)}")
        
        # Create and start the process
        try:
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=os.path.abspath(root),
                env=os.environ  # Pass through environment variables
            )
        except Exception as e:
            error_msg = f"Error: Failed to start aider process: {e}\n"
            logger.error(error_msg)
            return {
                "raw-stdout": "",
                "raw-stderr": error_msg
            }

        # Collect output
        if stream:
            async def event_generator():
                stdout_lines = []
                stderr_lines = []
                
                while True:
                    try:
                        # Read from stdout and stderr concurrently
                        stdout_data, stderr_data = await asyncio.gather(
                            process.stdout.readline(),
                            process.stderr.readline()
                        )
                        
                        if stdout_data:
                            line = stdout_data.decode().rstrip()
                            stdout_lines.append(line + "\n")
                            yield {
                                "event": "progress",
                                "data": json.dumps({"type": "stdout", "content": line})
                            }
                        if stderr_data:
                            line = stderr_data.decode().rstrip()
                            stderr_lines.append(line + "\n")
                            yield {
                                "event": "progress",
                                "data": json.dumps({"type": "stderr", "content": line})
                            }
                            
                        if not stdout_data and not stderr_data:
                            # Process complete, send final result
                            result = {
                                "raw-stdout": "".join(stdout_lines),
                                "raw-stderr": "".join(stderr_lines)
                            }
                            
                            if "https://aider.chat/docs/troubleshooting" in result["raw-stdout"]:
                                result["error"] = "something went wrong"
                                if "models-and-keys.html" in result["raw-stdout"]:
                                    result["error"] += ", AI key or model not found"
                            
                            yield {
                                "event": "complete",
                                "data": json.dumps(result)
                            }
                            break
                            
                    except Exception as e:
                        logger.error(f"Error processing output: {e}")
                        yield {
                            "event": "error",
                            "data": json.dumps({"error": f"Failed to process output: {e}"})
                        }
                        break

            return event_generator()
        else:
            # Non-streaming mode
            stdout_lines = []
            stderr_lines = []
            
            while True:
                try:
                    stdout_data, stderr_data = await asyncio.gather(
                        process.stdout.read(),
                        process.stderr.read()
                    )
                    
                    if stdout_data:
                        stdout_lines.extend(stdout_data.decode().splitlines(True))
                    if stderr_data:
                        stderr_lines.extend(stderr_data.decode().splitlines(True))
                        
                    if not stdout_data and not stderr_data:
                        break
                        
                except Exception as e:
                    logger.error(f"Error processing output: {e}")
                    stderr_lines.append(f"Error: Failed to process output: {e}\n")
                    break

            await process.wait()

            result = {
                "raw-stdout": "".join(stdout_lines),
                "raw-stderr": "".join(stderr_lines)
            }

            if "https://aider.chat/docs/troubleshooting" in result["raw-stdout"]:
                result["error"] = "something went wrong"
                if "models-and-keys.html" in result["raw-stdout"]:
                    result["error"] += ", AI key or model not found"

            return result

@app.post("/run-aider")
async def run_aider(request: AiderRequest):
    try:
        result = await collect_aider_output(
            request.message,
            request.files,
            request.auto_commits,
            request.dirty_commits,
            request.dry_run,
            request.root,
            request.stream
        )
        
        if request.stream:
            return EventSourceResponse(result)
        else:
            return JSONResponse(content=result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    import argparse
    import uvicorn
    
    # Set up argument parser
    parser = argparse.ArgumentParser(description='Run the Aider API server')
    parser.add_argument('--host', default='127.0.0.1', help='Host to listen on (default: 127.0.0.1)')
    parser.add_argument('--port', type=int, default=8000, help='Port to listen on (default: 8000)')
    
    args = parser.parse_args()
    uvicorn.run(app, host=args.host, port=args.port)

File: src/aider_api.ts
import fastify, { FastifyInstance } from 'fastify';
import { spawn } from 'child_process';
import path from 'path';
import fs from 'fs/promises';
import { FastifyRequest, FastifyReply } from 'fastify';

// Get the Node.js process
declare const process: {
  execPath: string;
  env: { [key: string]: string | undefined };
  argv: string[];
  exit: (code?: number) => never;
  cwd(): string;
};

// Configure logging
const logger = {
  debug: (msg: string) => {
    fs.appendFile('aider_api.log', `${new Date().toISOString()} - DEBUG - ${msg}\n`)
      .catch((err: Error) => console.error(err.message));
    console.debug(`${new Date().toISOString()} - DEBUG - ${msg}`);
  },
  error: (msg: string) => {
    fs.appendFile('aider_api.log', `${new Date().toISOString()} - ERROR - ${msg}\n`)
      .catch((err: Error) => console.error(err.message));
    console.error(`${new Date().toISOString()} - ERROR - ${msg}`);
  }
};

// Types
interface AiderRequest {
  message: string;
  files?: { [key: string]: string };
  auto_commits: boolean;
  dirty_commits: boolean;
  dry_run: boolean;
  root?: string;
}

interface AiderResponse {
  'raw-stdout': string;
  'raw-stderr': string;
  error?: string;
}

// Create Fastify instance
const app: FastifyInstance = fastify({
  logger: true
});

// Helper function to get aider executable path
function getAiderPath(): string {
  return path.join(process.cwd(), 'venv', 'bin', 'aider');
}

// Main function to collect aider output
async function collectAiderOutput(
  message: string,
  files: { [key: string]: string } | undefined,
  auto_commits: boolean,
  dirty_commits: boolean,
  dry_run: boolean,
  root: string = "."
): Promise<AiderResponse> {
  logger.debug(`Starting collectAiderOutput with message: ${message}`);
  logger.debug(`Files to process: ${files ? Object.keys(files) : 'None'}`);

  try {
    // Get aider executable path
    const aiderPath = getAiderPath();
    logger.debug(`Using aider from: ${aiderPath}`);

    // Prepare the aider command
    const cmd = [
      aiderPath,
      "--message", message,
      auto_commits ? "--auto-commits" : "--no-auto-commits",
      dirty_commits ? "--dirty-commits" : "--no-dirty-commits",
      dry_run ? "--dry-run" : "--no-dry-run",
      "--no-show-model-warnings",
      "--yes"
    ];

    // Add files if provided
    if (files) {
      cmd.push(...Object.keys(files));
    }

    logger.debug(`Working directory: ${path.resolve(root)}`);
    logger.debug(`Executing command: ${cmd.join(' ')}`);

    // Create and start the process
    const childProcess = spawn(cmd[0], cmd.slice(1), {
      cwd: path.resolve(root),
      env: process.env
    });

    let stdoutData = '';
    let stderrData = '';

    // Collect output
    childProcess.stdout.on('data', (data: Buffer) => {
      stdoutData += data.toString();
    });

    childProcess.stderr.on('data', (data: Buffer) => {
      stderrData += data.toString();
    });

    // Wait for process to complete
    const exitCode = await new Promise<number>((resolve) => {
      childProcess.on('close', resolve);
    });

    const response: AiderResponse = {
      'raw-stdout': stdoutData,
      'raw-stderr': stderrData
    };

    // Check for specific error conditions in stdout
    if (stdoutData.includes('https://aider.chat/docs/troubleshooting')) {
      response.error = 'something went wrong';
      
      if (stdoutData.includes('models-and-keys.html')) {
        response.error += ', AI key or model not found';
      }
    }

    return response;

  } catch (error) {
    logger.error(`Error: Failed to execute aider process: ${error instanceof Error ? error.message : String(error)}`);
    return {
      'raw-stdout': '',
      'raw-stderr': `Error: Failed to execute aider process: ${error instanceof Error ? error.message : String(error)}\n`
    };
  }
}

// Define the POST endpoint
app.post('/run-aider', async (request: FastifyRequest, reply: FastifyReply) => {
  try {
    const { message, files, auto_commits, dirty_commits, dry_run, root } = request.body as AiderRequest;
    const result = await collectAiderOutput(
      message,
      files,
      auto_commits,
      dirty_commits,
      dry_run,
      root
    );
    return result;
  } catch (error) {
    logger.error(`Error in /run-aider endpoint: ${error instanceof Error ? error.message : String(error)}`);
    reply.status(500).send({ error: error instanceof Error ? error.message : String(error) });
  }
});

// Main function to start the server
async function main() {
  try {
    const args = process.argv.slice(2);
    let host = '127.0.0.1';
    let port = 8000;

    // Parse command line arguments
    for (let i = 0; i < args.length; i++) {
      if (args[i] === '--host' && i + 1 < args.length) {
        host = args[i + 1];
        i++;
      } else if (args[i] === '--port' && i + 1 < args.length) {
        port = parseInt(args[i + 1], 10);
        i++;
      }
    }

    await app.listen({ host, port });
    console.log(`Server running at http://${host}:${port}`);
  } catch (err) {
    console.error('Error starting server:', err);
    process.exit(1);
  }
}

// Start the server if this is the main module
if (require.main === module) {
  main();
}

export default app;
